Basic Obstacle Avoidance

ErgiGU opened the pull request:
Implemented basic controls and auto-stop for the car(the car stops when it detects an obstacle that's within 0 and 150cm). This is just for the assignment and it will be improved further in the future. Also, fixed a problem where the car would wouldn't be able to go backwards as soon as it stops(because it's still within 0 and 150cm). The problem was fixed by adding a boolean.

Co-Authored-By: Yajiri <90855500+Yajiri@users.noreply.github.com>
Co-Authored-By: Aditya-Nair-gu <81224841+Aditya-Nair-gu@users.noreply.github.com>

ErgiGU committed:
Basic Obstacle Avoidance

Implemented basic controls and auto-stop for the car(the car stops when it detects an obstacle that's within 0 and 150cm). This is just for the assignment and it will be improved further in the future.

Co-Authored-By: Yajiri <90855500+Yajiri@users.noreply.github.com>
Co-Authored-By: Aditya-Nair-gu <81224841+Aditya-Nair-gu@users.noreply.github.com>

Yajiri reviewed (dismissed):
Looks great everything was working when we left.

platisd commented on arduino/smartcar/smartcar.ino:
```diff
@@ -1,9 +1,59 @@
+
+void autoStop(){
+  char input = Serial.read();
+  if(front.getDistance()>0 and front.getDistance()<150 and !flag){
```
What happens if in the first call to `getDistance` you get 200 and in the second 0? The car will stop while it shouldn't.

Yajiri replied:
That's impossible due to the limitations of the car in the module, it would need way more speed.
And in the if statement it's a dual condition (the reading needs to be 150>reading<0) otherwise it won't brake, or does this code work differently from Java? We ran it in the module and we never saw the problem you're telling us about.

platisd replied:
> That's impossible due to the limitations of the car in the module, it would need way more speed.

It's definitely possible. The measurements are noisy especially at such large distances. The fact that you didn't encounter the issue, doesn't mean it doesn't exist. Both the problem and the question remain:

> What happens if in the first call to getDistance you get 200 and in the second 0?

Yajiri replied:
Ahh we realized the problem but it's fun that we never saw this in the module when running it

platisd replied:
:+1:

ErgiGU replied:
Fixed it(you can check again)

ErgiGU committed:
if-statement improvement

Fixed the issue where it was getting the distance 2 times in the if-statement by getting it only once(by saving it in a variable before the if-statement)
